### PR TITLE
Add jacobian function

### PR DESCRIPTION
**Pull request by ChrisRackauckas**
Fixes https://github.com/FluxML/Zygote.jl/issues/98

### CODE REVIEW

File: src/lib/utils.jl [hunk 1]
```diff
@@ -99,6 +99,26 @@ macro showgrad(x)
     end)
 end
 
+"""
+jacobian(f,x)
+
+Construct the Jacobian of `f` where `x` is a real-valued array
+and `f(x)` is real-valued.
```
Comment (CarloLucibello) [inline src/lib/utils.jl:106]:
```suggestion
and `f(x)` is also a real-valued array.
```

**Comment by CarloLucibello** [comment]
needs a test and a mention in the docs

**Comment by cossio** [comment]
Bump?

**Comment by MasonProtter** [comment]
If we split out `OneHotVector` from Flux.jl, we could have this simpler implementation that's sometimes more performant and has less weird edge cases:
```julia
using Flux: OneHotVector

"""
jacobian(f,x)
Construct the Jacobian of `f` where `x` is a real-valued array
and `f(x)` is also a real-valued array.
"""
function jacobian(f, x)
    y, ∂f = Zygote.pullback(f,x)
    n, m  = length(x), length(y)
    mapreduce(i -> ∂f(OneHotVector(i, m)), hcat, 1:m)
end
```

File: src/lib/utils.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (ChrisRackauckas) [reply src/lib/utils.jl]: I'm not sure. I think people either know that or "want a Jacobian", in which case it's just an odd detail.

File: src/lib/utils.jl [hunk 2]
```diff
@@ -99,6 +99,26 @@ macro showgrad(x)
...
+
+Construct the Jacobian of `f` where `x` is a real-valued array
+and `f(x)` is also a real-valued array.
+"""
+function jacobian(f,x)
+    y,back  = Zygote.pullback(f,x)
+    k  = length(y)
+    n  = length(x)
+    J  = Matrix{eltype(y)}(undef,k,n)
+    e_i = fill!(similar(x), 0)
```
Comment (DhairyaLGandhi) [inline src/lib/utils.jl:113]: `zero(x)` would be clearer

**Review by DhairyaLGandhi — COMMENTED** [review]
Looks good, but needs tests

**Comment by DhairyaLGandhi** [comment]
Also probably something that should hook into the forward mode

**Comment by ChrisRackauckas** [comment]
Forward mode Jacobian should be kept as a separate call: which one is better depends on the application so both are required. What this should really connect to is SparseDiffTools for sparse Jacobians, but what will happen there is that SparseDiffTools will have a... more intense version that handles all of that extra machinery.

**Comment by DhairyaLGandhi** [comment]
This is due to it detecting sparsity patterns I assume?

File: src/lib/utils.jl (hunk 2 above)
Comment (mcabbott) [reply src/lib/utils.jl:113]: But not mutable -- see earlier suggestion.

Comment (MasonProtter) [reply src/lib/utils.jl:113]:
```suggestion
    e_i = fill!(similar(y), 0)
```
For reverse diff, your `e_i` is the wrong size unless your jacobian happens to be square.

**Review by MasonProtter — CHANGES_REQUESTED** [review]

**Comment by ChrisRackauckas** [comment]
>This is due to it detecting sparsity patterns I assume?

Well this could be made to accept a color vector even without SparseDiffTools.

File: src/lib/utils.jl (hunk 1 above)
[the remark this answers is not in the document]
Comment (mcabbott) [reply src/lib/utils.jl]: The point would be to remind them which way around the indices are, and that dimensions of x are flattened not kept. These are programming conventions not mathematical ones.

File: src/lib/utils.jl [hunk 3]
```diff
@@ -99,6 +99,26 @@ macro showgrad(x)
...
+    k  = length(y)
+    n  = length(x)
+    J  = Matrix{eltype(y)}(undef,k,n)
+    e_i = fill!(similar(y), 0)
+    @inbounds for i = 1:k
+        e_i[i] = oneunit(eltype(x))
+        J[i,:] = back(e_i)[1]
+        e_i[i] = zero(eltype(x))
+    end
+    J
```
Comment (mcabbott) [inline src/lib/utils.jl:119]:
```suggestion
    (J,)
```
Returning a 1-tuple allow for a consistent `jacobian(f, x, y)` without breaking anything, and would match what `gradient` does.

**Pull request closed without merging**